File: crates/xtriever-cli/src/wiki/rules.rs
//! Exclusion rules from the manifest, applied in order — first match wins and is the rule
//! counted (spec FR-003; research D3).

use serde::{Deserialize, Serialize};

/// One exclusion rule, as data.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(tag = "rule", rename_all = "snake_case")]
pub enum Rule {
    /// Exclude when the title ends with the value.
    TitleSuffix {
        /// The suffix.
        value: String,
    },
    /// Exclude when the value occurs within the first `within_chars` characters of the body.
    LeadContains {
        /// The phrase.
        value: String,
        /// The window, in characters (not bytes).
        within_chars: usize,
    },
}

/// The index of the first rule `title`/`text` match, if any.
#[must_use]
pub fn excluded_by(rules: &[Rule], title: &str, text: &str) -> Option<usize> {
    rules.iter().position(|rule| match rule {
        Rule::TitleSuffix { value } => title.ends_with(value.as_str()),
        Rule::LeadContains {
            value,
            within_chars,
        } => {
            // The phrase counts if it *starts* inside the window.
            let window_end = text
                .char_indices()
                .nth(*within_chars)
                .map_or(text.len(), |(i, _)| i);
            text.find(value.as_str()).is_some_and(|at| at < window_end)
        }
    })
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
mod tests {
    use super::*;

    fn rules() -> Vec<Rule> {
        vec![
            Rule::TitleSuffix {
                value: " (disambiguation)".to_owned(),
            },
            Rule::LeadContains {
                value: "may refer to".to_owned(),
                within_chars: 300,
            },
            Rule::LeadContains {
                value: "may mean".to_owned(),
                within_chars: 300,
            },
        ]
    }

    #[test]
    fn title_suffix_is_rule_zero() {
        assert_eq!(
            excluded_by(&rules(), "Flame (disambiguation)", "Flame may refer to:"),
            Some(0)
        );
        assert_eq!(excluded_by(&rules(), "Flame", "Fire is hot."), None);
    }

    #[test]
    fn lead_contains_counts_characters_not_bytes_and_stops_at_the_window() {
        let r = rules();
        let at_250 = format!("{}may refer to: things", "é".repeat(250));
        assert_eq!(
            excluded_by(&r, "X", &at_250),
            Some(1),
            "250 chars in (500 bytes) is inside 300 chars"
        );
        let at_350 = format!("{}may refer to: things", "a".repeat(350));
        assert_eq!(excluded_by(&r, "X", &at_350), None);
        assert_eq!(
            excluded_by(&r, "X", "Foo may mean several things."),
            Some(2)
        );
        // Straddling the window boundary counts if the phrase starts inside it.
        let straddle = format!("{}may refer to", "a".repeat(290));
        assert_eq!(excluded_by(&r, "X", &straddle), Some(1));
    }

    #[test]
    fn the_first_matching_rule_is_the_one_counted() {
        assert_eq!(
            excluded_by(&rules(), "Bank (disambiguation)", "Bank may mean:"),
            Some(0)
        );
        assert_eq!(
            excluded_by(&rules(), "Bank", "Bank may refer to: it may mean:"),
            Some(1)
        );
    }

    #[test]
    fn rules_round_trip_through_the_manifest_json() {
        let json = r#"[
            { "rule": "title_suffix", "value": " (disambiguation)" },
            { "rule": "lead_contains", "value": "may refer to", "within_chars": 300 }
        ]"#;
        let parsed: Vec<Rule> = serde_json::from_str(json).unwrap();
        assert_eq!(parsed[..], rules()[..2]);
    }
}

File: apps/python-wiki-demo/wikidemo/rules.py
"""The snapshot manifest's exclusion rules — `crates/xtriever-cli/src/wiki/rules.rs` in
Python (research D6): applied in order, the first match wins and is the rule counted.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Rule:
    kind: str  # "title_suffix" | "lead_contains"
    value: str
    within_chars: int | None = None


def parse_rules(exclusions: list[dict]) -> list[Rule]:
    rules = []
    for e in exclusions:
        if e["rule"] == "title_suffix":
            rules.append(Rule("title_suffix", e["value"]))
        elif e["rule"] == "lead_contains":
            rules.append(Rule("lead_contains", e["value"], int(e["within_chars"])))
        else:
            raise ValueError(f"unknown exclusion rule {e['rule']!r}")
    return rules


def excluded_by(rules: list[Rule], title: str, text: str) -> int | None:
    """The index of the first rule the article matches, or None.

    `lead_contains` counts if the phrase *starts* inside the first `within_chars` characters
    (characters, not bytes — Python string offsets are already characters)."""
    for i, rule in enumerate(rules):
        if rule.kind == "title_suffix":
            if title.endswith(rule.value):
                return i
        else:
            at = text.find(rule.value)
            if at != -1 and at < rule.within_chars:
                return i
    return None


def rule_name(rule: Rule) -> str:
    """The name the Rust build counts under (`build.rs`)."""
    if rule.kind == "title_suffix":
        return f"title_suffix:{rule.value}"
    return f"lead_contains:{rule.value}:{rule.within_chars}"
